File: src/go/problems/delete_middle_test.go
// https://leetcode.com/problems/delete-the-middle-node-of-a-linked-list/

package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type DeleteMiddle struct{}

func TestDeleteMiddle(t *testing.T) {
	gen := core.TestSuite[DeleteMiddle]{}
	gen.Add(func(tc *core.TestCase) {
		l1 := ListNode{
			Val: 1,
			Next: &ListNode{
				Val: 2,
				Next: &ListNode{
					Val: 3,
					Next: &ListNode{
						Val: 4,
					},
				},
			},
		}

		result := ListNode{
			Val: 1,
			Next: &ListNode{
				Val: 2,
				Next: &ListNode{
					Val: 4,
				},
			},
		}
		tc.Param(l1).Result(result)
	}).Run(t)
}

func (DeleteMiddle) Solution(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return nil
	}

	slow := head
	fast := head.Next.Next

	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	slow.Next = slow.Next.Next

	return head
}

File: src/python/problems/delete_middle_test.py
# https://leetcode.com/problems/delete-the-middle-node-of-a-linked-list/

from typing import Optional
from core.problem_base import *
from models.list_node import *

class DeleteMiddle(ProblemBase):
    def Solution(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head.next:
            return None

        slow = head
        fast = head.next.next

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        slow.next = slow.next.next

        return head


if __name__ == '__main__':
    TestGen(DeleteMiddle) \
        .Add(lambda tc: tc.Param("head", ListNode(1, ListNode(2, ListNode(3, ListNode(4))))).Result(ListNode(1, ListNode(2, ListNode(4))))) \
        .Run()